vm/systemSmartContracts/staking.go

```diff
@@ -88,16 +113,206 @@ func (r *stakingSC) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCod
+	if !bytes.Equal(args.CallerAddr, r.stakeAccessAddr) {
+		log.Debug("stake function not allowed to be called by", "address", args.CallerAddr)
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) < 2 {
+		return vmcommon.UserError
+	}
+
+	newRewardAddress := args.Arguments[0]
+	if len(newRewardAddress) != len(args.CallerAddr) {
+		return vmcommon.UserError
+	}
+
+	for i := 1; i < len(args.Arguments); i++ {
+		blsKey := args.Arguments[i]
+		stakedData, err := r.getOrCreateRegisteredData(blsKey)
+		if err != nil {
+			return vmcommon.UserError
+		}
+		if len(stakedData.RewardAddress) == 0 {
+			continue
+		}
+
+		stakedData.RewardAddress = newRewardAddress
+		err = r.saveStakingData(blsKey, stakedData)
+		if err != nil {
+			return vmcommon.UserError
+		}
+	}
+
+	return vmcommon.Ok
+}
+
 func (r *stakingSC) unJail(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, r.stakeAccessAddr) {
+		log.Debug("stake function not allowed to be called by", "address", args.CallerAddr)
+		return vmcommon.UserError
+	}
+
+	for _, argument := range args.Arguments {
+		stakedData, err := r.getOrCreateRegisteredData(argument)
+		if err != nil {
+			return vmcommon.UserError
+		}
+		if len(stakedData.RewardAddress) == 0 {
+			return vmcommon.UserError
+		}
+
+		stakedData.StakeValue = r.calculateStakeAfterBleed(
+			stakedData.JailedRound,
+			r.eei.BlockChainHook().CurrentRound(),
+			stakedData.StakeValue,
+		)
+		stakedData.JailedRound = math.MaxUint64
+
+		err = r.saveStakingData(argument, stakedData)
+		if err != nil {
+			return vmcommon.UserError
+		}
+	}
+
 	return vmcommon.Ok
 }
 
+func (r *stakingSC) getOrCreateRegisteredData(key []byte) (*StakedData, error) {
+	registrationData := StakedData{
+		RegisterNonce: 0,
+		Staked:        false,
+		UnStakedNonce: 0,
+		UnStakedEpoch: 0,
+		RewardAddress: nil,
+		StakeValue:    big.NewInt(0),
+		JailedRound:   math.MaxUint64,
+	}
+
+	data := r.eei.GetStorage(key)
+	if data != nil {
+		err := json.Unmarshal(data, &registrationData)
+		if err != nil {
+			log.Debug("unmarshal error on staking SC stake function",
+				"error", err.Error(),
+			)
+			return nil, err
+		}
+	}
+
+	return &registrationData, nil
+}
+
+func (r *stakingSC) saveStakingData(key []byte, stakedData *StakedData) error {
+	data, err := json.Marshal(*stakedData)
+	if err != nil {
+		log.Debug("marshal error on staking SC stake function ",
+			"error", err.Error(),
+		)
+		return err
+	}
+
+	r.eei.SetStorage(key, data)
+	return nil
+}
+
 func (r *stakingSC) jail(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, r.jailAccessAddr) {
+		return vmcommon.UserError
+	}
+
+	for _, argument := range args.Arguments {
+		stakedData, err := r.getOrCreateRegisteredData(argument)
+		if err != nil {
+			return vmcommon.UserError
+		}
+		if len(stakedData.RewardAddress) == 0 {
+			return vmcommon.UserError
+		}
+
+		stakedData.JailedRound = r.eei.BlockChainHook().CurrentRound()
```

Comment:
I presume this will come in effect (the peer will no longer be a validator/proposer) also at eoe

Reply:
EOE.